src/btree/bt_walk.c

```diff
@@ -414,24 +414,41 @@ __tree_walk_internal(WT_SESSION_IMPL *session,
 				empty_internal = false;
 			}
 
-			/*
-			 * Optionally return internal pages. Swap our previous
-			 * hazard pointer for the page we'll return. We don't
-			 * handle restart or not-found returns, it would require
-			 * additional complexity and is not a possible return:
-			 * we're moving to the parent of the current child page,
-			 * the parent can't have been evicted.
-			 */
-			if (!LF_ISSET(WT_READ_SKIP_INTL)) {
-				WT_ERR(__wt_page_swap(
-				    session, couple, ref, flags));
-				couple = NULL;
-				*refp = ref;
-				goto done;
-			}
-
 			/* Encourage races. */
 			__wt_timing_stress(session, WT_TIMING_STRESS_SPLIT_8);
+
+			/* Optionally return internal pages. */
+			if (LF_ISSET(WT_READ_SKIP_INTL))
+				continue;
+
+			for (;;) {
+				/*
+				 * Swap our previous hazard pointer for the page
+				 * we'll return.
+				 *
+				 * Not-found is an expected return, as eviction
+				 * might have been attempted. The page can't be
+				 * evicted, we're holding a hazard pointer on a
+				 * child, spin until we're successful.
+				 *
+				 * Restart is not expected, our parent WT_REF
+				 * should not have split.
+				 */
+				ret = __wt_page_swap(session,
+				    couple, ref, WT_READ_NOTFOUND_OK | flags);
+				if (ret == 0) {
+					/* Success, "couple" released. */
+					couple = NULL;
+					*refp = ref;
+					goto done;
+				}
+
+				 WT_ASSERT(session, ret == WT_NOTFOUND);
```

Comment:
Is this assertion supposed to stay long term?  Can't we get `WT_RESTART` or `EBUSY` here (at least)?

Reply:
Sorry, I see the comment about restart above: still curious if WT_NOTFOUND is really the only possible case.

Reply:
I don't think `WT_RESTART` should happen here (and I don't see the path where `EBUSY` matters, can you explain further what you're thinking?)

So, I was planning to leave the assert in, just so we know if it ever happens. I think the `WT_ERR_NOTFOUND_OK` macro should do the right thing in non-diagnostic builds, so we won't fail in the field.

Reply:
I'm also concerned about spurious failures in testing, but I'm happy to merge and see what happens.

Reply:
> I'm also concerned about spurious failures in testing

Agreed. I kind of want to see what happens, though, this problem could explain some weirdness in eviction because it could prevent us from evicting big parts of a busy tree.